### ISSUE TITLE
trouble reading interactions.tsv.gz into a Pandas dataframe

### ISSUE THREAD

**Issue by jhpoelen** [globalbioticinteractions/globalbioticinteractions#594]
from - https://github.com/globalbioticinteractions/globalbioticinteractions/issues/569#issuecomment-766695126 reported by @pwrose -

> I have trouble reading this file into a Pandas dataframe. It complains
about line 3626856:

> ParserError: Error tokenizing data. C error: Expected 84 fields in
line 3626856, saw 119 <https://zenodo.org/record/4460654/files/interactions.tsv.gz'%3C/span%3E%3Cspan>

**Comment by jhpoelen** [comment]
hi @pwrose - 

Unfortunately, I was unable to independently reproduce your claim that line 3626856 in https://zenodo.org/record/4460654/files/interactions.tsv.gz had 119 columns instead of 84.

According to the method below, each and every of the ~7.8M lines in the file had 84 fields separated by tabs.  

```shell
$ curl "https://zenodo.org/record/4460654/files/interactions.tsv.gz" | gunzip | awk -F '\t' '{ print NF }' | sort | uniq -c
7822671 84
``` 

I suspect that pandas, or some related program you are using is using methods other than just tabs to separate the fields. For instance, pandas might interpret quotes, commas or other non-tab characters as separators or escape characters. For some reason, many well-used program did not respect the IANA tsv definition (see https://www.iana.org/assignments/media-types/text/tab-separated-values and https://en.wikipedia.org/wiki/Tab-separated_values) .

A prominent example of incorrectly parsing of tsv files is https://github.com/tidyverse/readr/issues/844 , in which tsv files are not read using the IANA tsv specification by default, but instead treat quotes as special characters. 

Due to previous instances of these kinds of tsv parsing issues, interactions.csv.gz was also made available. For some reason, the more complicated csv format (e.g., escaping comma's) causes fewer parsing issues in commonly used libraries.

Hope this helps and curious to hear your thoughts.

**Comment by jhpoelen** [comment]
Just for completeness,  here's the reportedly offending line 3626856 including lines 3626855 and 3626857. Note that each line is prefixed by a line number. 
```
$ cat interactions.tsv.gz | gunzip | grep -n ".*" | tail -n+3626855 | head -n3
3626855:EOL:160225	ALATaxon:bb0578b2-06bb-4b7e-bf9c-2544a723d0cc | EOL:160225 | GBIF:5265721 | IF:199580 | IRMNG:11204282 | NCBI:1383253 | OTT:3684638 | WD:Q10510687	Synchytrium aureum	species	Fungi | Chytridiomycota | Chytridiomycetes | Chytridiales | Synchytriaceae | Synchytrium | Synchytrium aureum	EOL:5559 | EOL:5767 | EOL:5768 | EOL:5779 | EOL:5780 | EOL:19517 | EOL:160225	kingdom | phylum | class | order | family | genus | species	Synchytrium aureum	EOL:160225	Synchytrium	EOL:19517	Synchytriaceae	EOL:5780	Chytridiales	EOL:5779	Chytridiomycetes	EOL:5768	Chytridiomycota	EOL:5767	Fungi	EOL:5559	https://mycoportal.org/portal/collections/individual/index.php?occid=1751823		BPI 793938											hasHost	http://purl.obolibrary.org/obo/RO_0002454	GBIF:6300538	GBIF:6300538 | INAT_TAXON:55778 | ITIS:529540 | NCBI:98710 | OTT:4727349 | OTT:711401 | WD:Q24691127 | WORMS:1157548 | WORMS:1159730	Petasites frigidus palmatus	variety	Plantae | Tracheophyta | Magnoliopsida | Asterales | Asteraceae | Petasites | Petasites frigidus | Petasites frigidus palmatus	GBIF:6 | GBIF:7707728 | GBIF:220 | GBIF:414 | GBIF:3065 | GBIF:3088276 | GBIF:3088323 | GBIF:6300538	kingdom | phylum | class | order | family | genus | species | variety	Petasites frigidus	GBIF:3088323	Petasites	GBIF:3088276	Asteraceae	GBIF:3065	Asterales	GBIF:414	Magnoliopsida	GBIF:220	Tracheophyta	GBIF:7707728	Plantae	GBIF:6																	Ellison Bay		https://en.wiktionary.org/wiki/support	Mycology Collections Data Portal (MyCoPortal). 2020. https://mycoportal.org. Accessed at <https://github.com/globalbioticinteractions/mycoportal/archive/375a26416ad05b5397dc6e525b3d0f1df94e0cc2.zip> on 23 Jan 2021.			Mycology Collections Data Portal (MyCoPortal). 2020. https://mycoportal.org	globalbioticinteractions/mycoportal	https://github.com/globalbioticinteractions/mycoportal/archive/375a26416ad05b5397dc6e525b3d0f1df94e0cc2.zip		2021-01-23T01:52:13.410Z
3626856:EOL:160225	ALATaxon:bb0578b2-06bb-4b7e-bf9c-2544a723d0cc | EOL:160225 | GBIF:5265721 | IF:199580 | IRMNG:11204282 | NCBI:1383253 | OTT:3684638 | WD:Q10510687	Synchytrium aureum	species	Fungi | Chytridiomycota | Chytridiomycetes | Chytridiales | Synchytriaceae | Synchytrium | Synchytrium aureum	EOL:5559 | EOL:5767 | EOL:5768 | EOL:5779 | EOL:5780 | EOL:19517 | EOL:160225	kingdom | phylum | class | order | family | genus | species	Synchytrium aureum	EOL:160225	Synchytrium	EOL:19517	Synchytriaceae	EOL:5780	Chytridiales	EOL:5779	Chytridiomycetes	EOL:5768	Chytridiomycota	EOL:5767	Fungi	EOL:5559	https://mycoportal.org/portal/collections/individual/index.php?occid=1751822		BPI 793937											hasHost	http://purl.obolibrary.org/obo/RO_0002454	EOL:595565	EOL:595565 | GBIF:10331804 | GBIF:10629295 | GBIF:3088323 | GBIF:8250458 | GBIF:9310372 | INAT_TAXON:55780 | IRMNG:10978162 | ITIS:36054 | NCBI:56019 | OTT:896532 | WD:Q1997222 | WORMS:1098636 | WORMS:1131101 | WORMS:1141022 | doi:10.5281/zenodo.291971 | http://taxon-concept.plazi.org/id/Plantae/Petasites_frigidus_Linnaeus_1753 | http://treatment.plazi.org/id/DF12B55428FFDE03A2C898B222F0A57D	Petasites frigidus	species	Plantae | Tracheophyta | Magnoliopsida | Asterales | Asteraceae | Petasites | Petasites frigidus	EOL_V2:281 | EOL:4077 | EOL:283 | EOL:4205 | EOL:4206 | EOL:59514 | EOL:595565	kingdom | phylum | class | order | family | genus | species	Petasites frigidus	EOL:595565	Petasites	EOL:59514	Asteraceae	EOL:4206	Asterales	EOL:4205	Magnoliopsida	EOL:283	Tracheophyta	EOL:4077	Plantae	EOL_V2:281																	Jamtland, Aresjon		https://en.wiktionary.org/wiki/support	Mycology Collections Data Portal (MyCoPortal). 2020. https://mycoportal.org. Accessed at <https://github.com/globalbioticinteractions/mycoportal/archive/375a26416ad05b5397dc6e525b3d0f1df94e0cc2.zip> on 23 Jan 2021.			Mycology Collections Data Portal (MyCoPortal). 2020. https://mycoportal.org	globalbioticinteractions/mycoportal	https://github.com/globalbioticinteractions/mycoportal/archive/375a26416ad05b5397dc6e525b3d0f1df94e0cc2.zip		2021-01-23T01:52:13.410Z
3626857:EOL:160225	ALATaxon:bb0578b2-06bb-4b7e-bf9c-2544a723d0cc | EOL:160225 | GBIF:5265721 | IF:199580 | IRMNG:11204282 | NCBI:1383253 | OTT:3684638 | WD:Q10510687	Synchytrium aureum	species	Fungi | Chytridiomycota | Chytridiomycetes | Chytridiales | Synchytriaceae | Synchytrium | Synchytrium aureum	EOL:5559 | EOL:5767 | EOL:5768 | EOL:5779 | EOL:5780 | EOL:19517 | EOL:160225	kingdom | phylum | class | order | family | genus | species	Synchytrium aureum	EOL:160225	Synchytrium	EOL:19517	Synchytriaceae	EOL:5780	Chytridiales	EOL:5779	Chytridiomycetes	EOL:5768	Chytridiomycota	EOL:5767	Fungi	EOL:5559	https://mycoportal.org/portal/collections/individual/index.php?occid=1751821		BPI 793936											hasHost	http://purl.obolibrary.org/obo/RO_0002454	EOL:595565	EOL:595565 | GBIF:10331804 | GBIF:10629295 | GBIF:3088323 | GBIF:8250458 | GBIF:9310372 | INAT_TAXON:55780 | IRMNG:10978162 | ITIS:36054 | NCBI:56019 | OTT:896532 | WD:Q1997222 | WORMS:1098636 | WORMS:1131101 | WORMS:1141022 | doi:10.5281/zenodo.291971 | http://taxon-concept.plazi.org/id/Plantae/Petasites_frigidus_Linnaeus_1753 | http://treatment.plazi.org/id/DF12B55428FFDE03A2C898B222F0A57D	Petasites frigidus	species	Plantae | Tracheophyta | Magnoliopsida | Asterales | Asteraceae | Petasites | Petasites frigidus	EOL_V2:281 | EOL:4077 | EOL:283 | EOL:4205 | EOL:4206 | EOL:59514 | EOL:595565	kingdom | phylum | class | order | family | genus | species	Petasites frigidus	EOL:595565	Petasites	EOL:59514	Asteraceae	EOL:4206	Asterales	EOL:4205	Magnoliopsida	EOL:283	Tracheophyta	EOL:4077	Plantae	EOL_V2:281																	Jamtland, Aresjon		https://en.wiktionary.org/wiki/support	Mycology Collections Data Portal (MyCoPortal). 2020. https://mycoportal.org. Accessed at <https://github.com/globalbioticinteractions/mycoportal/archive/375a26416ad05b5397dc6e525b3d0f1df94e0cc2.zip> on 23 Jan 2021.			Mycology Collections Data Portal (MyCoPortal). 2020. https://mycoportal.org	globalbioticinteractions/mycoportal	https://github.com/globalbioticinteractions/mycoportal/archive/375a26416ad05b5397dc6e525b3d0f1df94e0cc2.zip		2021-01-23T01:52:13.410Z
```

**Comment by jhpoelen** [comment]
Note that line 3626856 contains a field with value ```Jamtland, Aresjon``` . In csv, this value with the embedded comma would have to be quotes, but in tsv, comma's need no special treatment.

**Comment by pwrose** [comment]
The .csv file works!

**Comment by jhpoelen** [comment]
@pwrose glad to hear that the csv file worked! Hurray!

If you have a moment, I'd be great if you could share a code snippet that succeeded reading the csv file and the code snippet that failed to read the tsv file. This way I hope to make sure that we pinpoint the issue to a suspected misconfigured tsv parser.

**Comment by jhpoelen** [comment]
@pwrose thanks for sharing your code snippets. Your tsv reading example confirm my suspicions in that tab-separated-files are often mistaken for comma-separated-file with a tab (i.e. ```\t```) as a separator. However, tsv files, according to IANA definition, do not have the complicated and confusing quoting and escaping that come with the csv files. 

In this case, the csv reader got confused by reading a tsv files because some character was misinterpreted as an unescaped control character. 

So, pandas is trying to be too smart and that causes problems. You can probably read valid tsv files by making sure that all the quoting etc. is disabled or by using a simple line-by-line string splitter. But hey, csv works, so it sounds like you a good to go!

In short - interactions.tsv.gz is valid IANA tsv , so I am closing this issue.  

Thanks for your patience!
